fixed: Add security headers, fix CSRF exemptions, and harden configuration

Pritz395 opened the pull request:
Implements comprehensive security hardening addressing Security Headers and CSRF Exempt Review.

CHANGES SUMMARY:

Part 1: Security Headers (blt/settings.py)
- Added HSTS with 1-year max-age, includeSubDomains, and preload
- Enabled SECURE_CONTENT_TYPE_NOSNIFF to prevent MIME type sniffing
- Enabled SECURE_BROWSER_XSS_FILTER for browser XSS protection
- Set SECURE_REFERRER_POLICY to "same-origin" to prevent referrer leakage
- Changed X_FRAME_OPTIONS from "SAMEORIGIN" to "DENY" for stricter clickjacking protection

Part 2: CSRF Exempt Review
CRITICAL FIX: Removed @csrf_exempt from set_theme() endpoint which had no authentication
- Added input validation to only allow "light" or "dark" theme values
- Improved error handling for JSON decode errors
- Changed error response from 400 to 405 for non-POST requests

Removed unnecessary @csrf_exempt from authenticated endpoints:
- resolve() in website/views/issue.py (already protected by @login_required)
- page_vote() in website/views/issue.py (already protected by @login_required)
- update_submission_status() in website/views/bitcoin.py (already protected by @login_required)

Note: Legitimate webhook endpoints remain unchanged (GitHub, Slack, Bounty webhooks)

Part 3: Additional Hardening
- Enhanced SECRET_KEY handling to require environment variable in production
- Added CSP configuration (requires django-csp package installation)
- Production will fail to start if SECRET_KEY is not set

FILES CHANGED:
- blt/settings.py - Security headers, CSP settings, SECRET_KEY hardening, X_FRAME_OPTIONS
- website/views/core.py - Removed @csrf_exempt from set_theme(), added input validation
- website/views/issue.py - Removed @csrf_exempt from resolve() and page_vote()
- website/views/bitcoin.py - Removed @csrf_exempt from update_submission_status()

Fixes #5370

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Security Enhancements**
  * Enforced non-default SECRET_KEY in production, added CSP and security headers/middleware, strengthened cookie and frame protections, and enabled CSRF on multiple endpoints.
* **API & Rate Limits**
  * Added anonymous and authenticated request throttling for the API.
* **Client behavior**
  * Theme persistence and interactive actions now source CSRF token from a meta tag and use CSRF-protected POSTs; client error handling adjusted.
* **Tests**
  * Updated tests to expect 405 Method Not Allowed for theme GET.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Pritz395 committed:
Add session security flags, fix timing attack vulnerability, and enable API rate limiting

Pritz395 committed:
Add ACME_KEY validation and security logging for failed token attempts

Pritz395 committed:
security: Add security headers, fix CSRF exemptions, and harden configuration

Pritz395 committed:
fix: Set CSRF_COOKIE_HTTPONLY to False and remove CSP/deprecated settings

Pritz395 committed:
fix: Read CSRF token from meta tag in darkMode.js and remove unused import

Pritz395 committed:
fix: Allow management commands to run without SECRET_KEY for CI/CD

Pritz395 committed:
fix: Remove redundant import and add more safe management commands

Pritz395 committed:
fix: Remove console.error statement from darkMode.js

Pritz395 committed:
fix: Update test to expect 405 instead of 400 for invalid method

Pritz395 committed:
fix: Complete security hardening per PR scope

- Fixed SECRET_KEY enforcement (removed backdoor, simplified logic)
- Added complete CSP configuration with django-csp
- Reverted throttle rate changes from /hour to /day (out of scope)
- Changed X_FRAME_OPTIONS to SAMEORIGIN (safer default)
- Added CSP middleware and app configuration

gojo-satorou-v7 commented:
Too many security headers does not guarantee security, it would be better if you included their use and for csp if you posted a video that everything is working correctly then that would be great.

Pritz395 committed:
chore: Add django-csp dependency for CSP headers support

Pritz395 committed:
chore: Update django-csp to version 4.0 and regenerate lockfile

Pritz395 committed:
fix: Add @ensure_csrf_cookie to set_theme and update CSP to django-csp 4.0 format

- Add @ensure_csrf_cookie decorator to set_theme() for unauthenticated users
- Update CSP configuration from CSP_* variables to CONTENT_SECURITY_POLICY dict format (django-csp 4.0)

Pritz395 committed:
fix: Update CSP to include unpkg, YouTube/Vimeo, remove HSTS preload

- Add unpkg.com to script-src and style-src (used for alpinejs, leaflet, tailwindcss)
- Add YouTube and Vimeo to frame-src for education videos
- Remove redundant bhfiles.storage.googleapis.com from img-src (covered by https:)
- Remove HSTS preload (not recommended per hstspreload.org)

Pritz395 committed:
Merge branch 'main' into security/add-headers-fix-csrf-5370

Pritz395 committed:
docs: Add comprehensive security headers documentation with attack prevention details

Pritz395 committed:
fix: Add missing CDN domains to CSP and improve security docs

- Added cdn.tailwindcss.com and cdnjs.cloudflare.com to CSP (used in email templates)
- Tightened img-src policy (explicit whitelist instead of permissive https:)
- Fixed HSTS preload documentation (removed misleading 'not recommended' note)
- Improved nosniff documentation (defense-in-depth context)
- Added comprehensive CSP testing and deployment instructions

Pritz395 committed:
style: Apply ruff-format auto-fixes (remove trailing blank lines)

Pritz395 committed:
fix: Change resolve() view to POST and add CSRF protection

- Added @require_POST decorator to resolve() view (state-changing operation should use POST)
- Updated resolveIssue() JavaScript to use POST method with CSRF token
- Fixed security issue: GET requests for state changes are vulnerable to CSRF attacks

gojo-satorou-v7 reviewed (changes requested):
You're still missing a lot of domains for example:

img-src:
<img width="972" height="434" alt="Image" src="https://github.com/user-attachments/assets/5019c574-9903-4e01-ac7a-861229ee5d39" />

script-src: 

<img width="978" height="400" alt="Image" src="https://github.com/user-attachments/assets/ba982f6f-13fc-4b9a-a180-319c97cca23a" />

I haven't listed all of them but I believe these are important enough to assume that the current csp implementation will break things.

Also since this implementation is going directly to prod I do not recommend implementing without extensive testing and what you can do instead is make this report-only (which you also mentioned in the comments).

Pritz395 committed:
Merge branch 'main' into security/add-headers-fix-csrf-5370
